iib/web/models.py

```diff
@@ -382,6 +392,64 @@ class Batch(db.Model):
 
     requests = db.relationship('Request', foreign_keys=[Request.batch_id], back_populates='batch')
 
+    @property
+    def state(self):
+        """
+        Get the state of the batch.
+
+        If one or more requests in the batch are ``in_progress``, then the batch is ``in_progress``.
+        Once all the requests in the batch have completed, if one or more requests are in the
+        ``failed`` state, then so is the batch. If all requests in the batch are in the ``complete``
+        state, then so is the batch.
+
+        :return: the state of the batch
+        :rtype: str
+        """
+        contains_failure = False
+        for state in self.request_states:
+            # If one of the requests is still in progress, the batch is also
+            if state == 'in_progress':
+                return 'in_progress'
+            elif state == 'failed':
+                contains_failure = True
+
+        # At this point, we know the batch is done
+        if contains_failure:
+            return 'failed'
+        else:
+            return 'complete'
+
+    @property
+    def request_ids(self):
+        """
+        Get the IDs of the requests in the batch.
+
+        :return: the set of requests IDs
+        :rtype: set
+        """
+        return {
+            result[0] for result in db.session.query(Request.id).filter(Request.batch == self).all()
```

Comment:
I think `results` is a named tuple. It might be more intuitive to use `result.id` instead of `result[0]`.